Katello/smart_proxy_container_gateway
# Fixes #31485 - Unauthenticated repo cache blocks access to repositories that require authentication

# To test:

0) Set up a smart proxy with Pulp 3.  You can use the following installer options to install Pulp 3:
```
--foreman-proxy-plugin-pulp-pulpcore-enabled 1
--foreman-proxy-plugin-pulp-pulpcore-mirror 1
```
You might hit some bumps with the installation, so let me know if you're stuck.

1) Clone the smart-proxy repo and copy over all config settings from the system-installed foreman-proxy.  You'll have to copy the certs to vagrant's home folder so it can access them and update the config files accordingly.
2) Once your smart proxy is working, install this smart proxy plugin.  You can do this by cloning the repo, adding its gem to `smart-proxy/bundler.d/Gemfile.local./rb`, and then copying the sample config file from this smart proxy plugin to `smart-proxy/config/settings.d/container_gateway.yml`.
3) Create a database called `smart_proxy_container_gateway` in postgres.  You'll need to make sure the database user you set in the config can read and write to this database.
4) Sync a docker repo to your smart proxy
5) Look at your catalog:  ```sudo podman search `hostname`/```
    - This should be empty
6) Find the name of your docker repo from your Katello server.  It'll look like `default_organization-environment-content_view-product-image`
7) Try a pull that should fail "unauthorized": ```sudo podman pull `hostname`/<docker repo name>```
8) Add your docker repo to the unauthenticated repo list via `httpie`:
```
http PUT https://`hostname`/v2/unauthenticated_repository_list repositories:='["<docker repo name>"]'
```
9) View your catalog again. It should have the repo you added
10) Try to `podman pull` the docker repo again. It should work now.

The idea is that, at smart proxy sync time, Katello will send a list of unauthenticated repos to the smart proxy via that unauthenticated_repository_list endpoint.

# Alternative testing
Use [this Katello PR](https://github.com/Katello/katello/pull/9043) to automatically update the unauthenticated repo list at smart proxy sync time.
